Track application state in URL

huddlej opened the pull request:
Adds support for tracking the state of the application's dropdowns and selected sites in the URL. Also adds a form field where users can manually specify sites to select. This form field will track sites selected by other interactions as well (e.g., brushing).

Closes #4

huddlej committed:
Track state of dropdown fields in the URL

Adds two new functions to set the state of the dropdowns from the given URL and
also to update the URL when the state of the dropdowns changes.

Also removes a call to "updateChart" from the initial site plot creation and
uses a call to "dropdownChange" instead to get the same effect based on the
state of the dropdowns. This may not be an appropriate solution and needs to be
tested more.

huddlej committed:
Update multiple fields at once

huddlej committed:
Do not allow users to specify invalid field values in the app state

Prevents bugs when switching between datasets when the URL requests a specific
value for a dropdown (e.g., condition or site metric) that doesn't exist for the
new dataset. The app now issues a warning on the console. This should be a
warning through a unified feedback mechanism in the app, though.

TODO: Make a unified feedback mechanism for the app.

huddlej committed:
Rename 'site' field to 'site_metric'

This commit aligns the naming conventions of the site metric dropdown with the
mutation metric dropdown and allows us to add a separate 'site' field later for
tracking selected sites.

huddlej committed:
Refactor logic to select/deselect sites into a separate function

Adds a new function `updateSites` that takes lists of sites to be selected
and/or deselected, selects and deselects those sites, and then updates the logo
plot. Updates all individual calls to selectSite and deselectSite to use this
new function, providing a unified interface for these updates and only exposing
this one function from the line plot's object.

This new function can be used to handle updates to the URL when tracking sites
in the URL and any other similar logic that happens every time the selected
sites change.

Closes #100.

huddlej committed:
Track selected sites in URL and form field

Adds ability to specify selected sites in the URL and in a form field near the
"clear selections" button. Selecting sites by clicking or brushing, etc. updates
the form field accordingly and changing the contents of the field manually
updates the sites that are selected. If sites are specified in the URL at load,
those sites are selected instead of the site with the maximum value. As selected
sites change, the URL is updated, too.

Although this implementation works, it feels gross. There is a lot of reaching
between main and line_plot_zoom js files using the DOM and I had to add a
promise to get things to work. There must be a better way to do this kind of
thing and the answer might be something like React et al.

huddlej committed:
Provide hints about valid inputs in form field placeholder text

skhilton reviewed (approved):
@huddlej this looks great! I am so excited to go make URLs for specific figures in Juhye's paper. 

There is one maybe bug but I don't think it is worth holding up the PR if it is not easy to fix. If I clear all selections and then paste the URL into a new tab, the top site is selected because it views this the initial load. This is definitely a corner case so it might not be worth it. I'll leave it and the merge up to you.

huddlej committed:
Only select max site when no selected sites are passed by URL

Allows the user to pass an empty list of selected sites in the URL such that
opening that URL will cause no sites to be selected. This fixes an edge case bug
where an empty string in the selected sites URL parameter was ignored and the
site with the maximum value was selected instead.

huddlej commented:
Thank you for catching that bug, @skhilton! I just added a fix for it. I'll merge this now.
